### ISSUE TITLE
prop-types false positive in inner block arrow function

### ISSUE THREAD

**Issue by marlonmleite** [yannickcr/eslint-plugin-react#2458]
**Component with arrow fn:**
```jsx
import React from 'react'

const ListCustomer = () => {
  const renderItem = ({ customer }) => (
    <Customer name={customer.name} />
  )

  return (
    <List renderItem={renderItem} />
  )
}

export default ListCustomer
```

**Component with fn:**
```jsx
import React from 'react'

function ListCustomer() {
  const renderItem = ({ customer }) => (
    <Customer name={customer.name} />
  )

  return (
    <List renderItem={renderItem} />
  )
}

export default ListCustomer
```

**This print a error:**
```sh
error  'customer.name' is missing in props validation  react/prop-types
```

In the 2 examples, you can never display the error.

**Comment by ljharb** [comment]
`renderItem` is a function that takes props and returns jsx, so it's a component.

The bug is that it's not warning in both cases.

**Comment by pawelnvk** [comment]
This seems to be more complex topic than assumed. Saying that component is a function that takes props and return jsx isn't 100% correct. 
1. There are some functions that don't accept props, yet we can call them functions:
```jsx
const Logo = () => <img alt="Company logo" src="/path/to/logo.svg" />
```
2. There are functions that don't return jsx, but still we should consider them components:
```jsx
const VisitTracker = ({ path }) => {
  useEffect(() => trackVisit(path), [path]);

  return null;
}
```
3. There could be even functions that can not take any props and return no jsx, yet they could be considered as component:
```jsx
const VisitTracker = () => {
  useEffect(() => {
    // Some interaction with DOM
  });

  return null;
}
```
I'm sure code above can be arguable, but it's more or less valid. Our assumption would mean that we could consider a lot of simple functions as component.

IMO what makes components a component is context how we use it. Render prop looks like a component but there is a big difference it doesn't take props, it takes arguments. For sake of validation cases described above I would say that we could use following pattern.
E.g.
```jsx
import React from 'react'

const ListCustomer = () => {
  const renderItem = ({ customer }) => (
    <Customer name={customer.name} />
  )

  return (
    <List renderItem={renderItem} />
  )
}

export default ListCustomer
```
1. Search for each function that might be component. (In this case `ListCustomer`, `renderItem`.)
2. Check if found function has higher in scope any other function. (In this case `ListCustomer` is the top level function, returns JSX so it should be considered a component, `renderItem` is in scope of `ListCustomer`. Since declaring components inside of render function of other component would be considered as huge antipatern we could stop right there, but for out sake we can go with that further.)
3. Check ho given function is used. If it's used as component in current scope, it is a component, if it's passed as prop it's render prop (In this case `renderItem` it's passed as prop.)

Since declaring render props outside of component (top level) seems to be not that good idea either. Such render props miss their point and we should consider them components. If someone insist on treating them as components we could handle adding some flag above function definition. (Comment approach is used in vue lint as well, if I'm correct)

**Comment by ljharb** [comment]
Our rules share logic for component detection. You’re right that it can be improved, but there’s also limits to static analysis.

If you want to make a PR to improve the accuracy of component detection, that’d be appreciated - if that incidentally fixes this bug, so be it :-) otherwise, this specific bug should be fixed so that both of these warn the same.

**Comment by golopot** [comment]
The two examples both warns correctly, in the latest version. I think the OP expected both of the examples to not warn.

**Comment by ljharb** [comment]
Then while improved component detection is always welcome, it seems like every thing is working correctly at the moment, so, closing.
